src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -833,6 +833,43 @@ void ObjectCopyRequest<I>::compute_dst_object_may_exist() {
                    << dendl;
 }
 
+template <typename I>
+void ObjectCopyRequest<I>::trigger_copyup() {
+  // With mirroring, a write to a non-existent primary cloned image object will 
+  // trigger a copyup which updates the older synced snaps with the parent data.
+  // This causes the rbd-mirror to ignore the parent data and only sync the new write
+  // to the clone, causing data corruption.
+  // For rbd-mirror scenarios, we trigger a copyup on the object and write the parent
+  // data to the object before writing the new clone image data.
+  if (m_src_snap_id_start != 0 && m_src_image_ctx->parent != nullptr) {
+    ldout(m_cct, 20) << dendl;
+    auto io_context = m_dst_image_ctx->duplicate_data_io_context();
+    auto ctx = create_context_callback<
+      ObjectCopyRequest<I>, &ObjectCopyRequest<I>::handle_trigger_copyup>(this);
+    if (io::util::trigger_copyup(
+      m_dst_image_ctx, m_dst_object_number, io_context, ctx)) {
+      return;
+    }
```

Comment:
(Not a review, just something that came up in a conversation about mocking)

This would hang (and leak some memory) if `io::util::trigger_copyup()` returns false because you just bail from `ObjectCopyRequest::trigger_copyup()` in that case.  If it's "not possible" for `io::util::trigger_copyup()` to return false here, an assert would be nice.